tokio-threadpool/src/deque.rs

```diff
@@ -0,0 +1,212 @@
+use std::mem::forget;
+use std::ptr;
+
+use std::sync::atomic::{AtomicIsize, AtomicPtr, fence};
+use std::sync::atomic::Ordering::{SeqCst, Acquire, Release, Relaxed};
+
+// Initial size for a buffer.
+// TODO: Much too small...
+static MIN_SIZE: usize = 32;
+
+#[derive(Debug)]
+pub struct Deque<T: Send> {
+    bottom: AtomicIsize,
+    top: AtomicIsize,
+    array: AtomicPtr<Buffer<T>>,
+}
+
+/// When stealing some data, this is an enumeration of the possible outcomes.
+#[derive(Debug, PartialEq)]
+pub enum Poll<T> {
+    /// The deque was empty at the time of stealing
+    Empty,
+    /// The stealer lost the race for stealing data, and a retry may return more
+    /// data.
+    Inconsistent,
+    /// The stealer has successfully stolen some data.
+    Data(T),
+}
+
+/// An internal buffer used by the chase-lev deque. This structure is actually
+/// implemented as a circular buffer, and is used as the intermediate storage of
+/// the data in the deque.
+///
+/// This type is implemented with *T instead of Vec<T> for two reasons:
+///
+///   1. There is nothing safe about using this buffer. This easily allows the
+///      same value to be read twice in to rust, and there is nothing to
+///      prevent this. The usage by the deque must ensure that one of the
+///      values is forgotten. Furthermore, we only ever want to manually run
+///      destructors for values in this buffer (on drop) because the bounds
+///      are defined by the deque it's owned by.
+///
+///   2. We can certainly avoid bounds checks using *T instead of Vec<T>, although
+///      LLVM is probably pretty good at doing this already.
+///
+/// Note that we keep old buffers around after growing because stealers may still
+/// be concurrently accessing them. The buffers are kept in a linked list, with
+/// each buffer pointing to the previous, smaller buffer. This doesn't leak any
+/// memory because all buffers in the list are freed when the deque is dropped.
+#[derive(Debug)]
+struct Buffer<T: Send> {
+    storage: *mut T,
+    size: usize,
+    prev: Option<Box<Buffer<T>>>,
+}
+
+impl<T: Send> Deque<T> {
+    pub fn new() -> Deque<T> {
+        let buf = Box::new(unsafe { Buffer::new(MIN_SIZE) });
+        Deque {
+            bottom: AtomicIsize::new(0),
+            top: AtomicIsize::new(0),
+            array: AtomicPtr::new(Box::into_raw(buf)),
+        }
+    }
+
+    #[inline]
+    pub unsafe fn push(&self, data: T) {
+        let b = self.bottom.load(Relaxed);
+        let t = self.top.load(Acquire);
+        let mut a = self.array.load(Relaxed);
+
+        // Grow the buffer if it is full.
+        let size = b.wrapping_sub(t);
+        if size == (*a).size() {
+            a = Box::into_raw(Box::from_raw(a).grow(b, t));
+            self.array.store(a, Release);
+        }
+
+        (*a).put(b, data);
+        fence(Release);
+        self.bottom.store(b.wrapping_add(1), Relaxed);
+    }
+
+    #[inline]
+    pub fn poll(&self) -> Poll<T> {
+        unsafe {
```

Comment:
Is this whole function actually unsafe? ISTM that the scope could be narrowed.

Reply:
Ah, thanks. This entire file is actually dead code right now :) The deque is provided by a lib.